TheJumpCloud/support
# Update Deploy-UserCertificate Functions to generate certs on SHA match

## Issues
* [CUT-4309](https://jumpcloud.atlassian.net/browse/CUT-4309) - Update Deploy-UserCertificate Functions to generate certs on SHA match

## What does this solve?

This release aims to reduce the amount of time required to generate and deploy certificates for users. In previous iterations of this script, the tool would generate a new command each time the script was run for each user. This version changes the command "trigger" to the SHA1 hash of a user's certificate and compares that value when running Start-DeployUserCerts.

This release also introduces a multi-threaded operation when the tool generates commands for "new" or "all" users in the radiusMembers list. 

## Is there anything particularly tricky?

## How should this be tested?

Functionality for the changes should be already covered in our Pester tests for this module. To validate the changes in this pull request. Pull the branch, open up the `scripts/automation/Radius` directory in a PowerShell session and run the pester tests on one of our Pester Tests orgs: 

./tests/Invoke-Pester.ps1 -JumpCloudApiKey {YourPesterTestAPIKey}

![Screenshot 2024-10-21 at 1 35 57 PM](https://github.com/user-attachments/assets/71a31b2d-1abe-453c-8160-73feb65448ea)

In terms of updating from a previous version of the 2.0.0 development branch of radius certs:

- simply running this new version will generate new cert commands for each user, these new commands will have a trigger containing the SHA1 value of the user's cert. Subsequent times the script is run, commands for those users will not explicitly be re-generated unless the `$forceGenerateCommands` parameter is specified.

Before:
![Screenshot 2024-10-11 at 1 56 47 PM](https://github.com/user-attachments/assets/91e33b18-cfe4-4812-8b2c-c0d9e527fb36)

After:
![Screenshot 2024-10-11 at 1 57 00 PM](https://github.com/user-attachments/assets/4a83f5e4-435c-43ed-9c7a-a9aeb72bc8e9)

## Screenshots

Preliminary testing indicates that this version of the script runs about 47% faster than the previous version. 

Newest version (running threads in parallel to generate commands)
![Screenshot 2024-10-21 at 1 16 11 PM](https://github.com/user-attachments/assets/ccff29b3-25c4-4247-a344-7bc676d2865e)
Previous version (running single threaded and in sequence to generate commands)
![Screenshot 2024-10-21 at 1 34 49 PM](https://github.com/user-attachments/assets/441c72be-a17f-4089-ae78-e0d5eba5dc68)




[CUT-4309]: https://jumpcloud.atlassian.net/browse/CUT-4309?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ
